### ISSUE TITLE
[Rust] Implement new Concept Exercise: vec-macro

### ISSUE THREAD

**Issue by workingjubilee** [exercism/rust#1063]
This issue describes how to implement the vec concept exercise for the Rust track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The vec collection is the most common data container in Rust, and as such has many miscellaneous functions. This should serve as an introduction to common usage of Vecs, and is not intended to cover its usage exhaustively. It should mention that Vec is not an array, but it is not required to treat arrays in detail and focus on Vec should be favored over going down a rabbithole.

## Learning objectives

Student should be able to...
- use `vec!`
- call basic methods on Vec
- recognize that Vec is not an array
- recognize the most obvious cases where Vec may reallocate
- recognize a few good use-cases for Vec

Student may be able to...
- use a Vec as the basis of an Iterator
- understand that Vec<T> is generic over T
- understand details of Vec's relationship with arrays
- take slices out of a Vec

## Out of scope

- complex data structures that may be formed with Vecs
- other std::collections, even VecDeque!
- algorithms for use with Vec, but do use them if they are both basic enough and it seems appropriate for teaching the data structure

## Concepts

- vec
- arrays?
- slices?

## Prerequisites

This exercise should depend only on a basic understanding of functions, types, and variables. They need to know enough to be able to shove something into a Vec, after all.

There are some uncertainties in placement, because this is both a suitable lesson to serve as a stepping stone to these concepts but also it is useful to (re)visit this after they have been understood.
- Is it before or after methods?
- Is it before or after Iterator?
- Is it before or after generics?

The lesson may change accordingly. Including variances is not required: assume the most minimal form. Additional content for possible usage may be included in the `.meta` folder of the exercise.

### Hints

- std::vec: https://doc.rust-lang.org/std/vec/index.html

### After

- Rust by Example (stdlib types): https://doc.rust-lang.org/rust-by-example/std/vec.html#vectors
- The Rust Programming Language (Creating a Vector): https://doc.rust-lang.org/book/ch08-01-vectors.html#creating-a-new-vector

## Representer

This should require no special changes to the representer.

## Analyzer

This should require no special changes to the analyzer.

## Implementing

See our [implementation guide](https://github.com/exercism/v3/blob/master/languages/rust/howto/implementing-a-concept-exercise.md#how-to-implement-a-rust-concept-exercise) for guidance. In addition you may want to make use of the shell script while in the `concept` directory, with the arg vec, e.g. `sh boil_exercise.sh vec`

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

**Comment by efx** [comment]
I'm renaming this issue to track finishing vec-macro.
See our discussion [here](https://github.com/exercism/v3/pull/2918#issuecomment-756743573) for the reasoning.

**Comment by coriolinus** [comment]
If you rename this, please create issues for the other three proposed `vec` concept exercises as well, so they don't get lost.

**Comment by efx** [comment]
#1103 #1104 #1105 capture the additional concept exercises.
